File: cplusplus/ext5/VisitorWithMood.cpp
#include "VisitorWithMood.hpp"

VisitorWithMood::VisitorWithMood(CoffeeHouse* model) : Visitor(model) {
	if (this->ifWantsToLeave())
		this->leaveQueue();
}

void VisitorWithMood::process(long currentTime)
{
	if (this->ifWantsToLeave())
		this->leaveQueue();
	else
		Visitor::process(currentTime);
}

bool VisitorWithMood::ifWantsToLeave()
{
	int mood = rand() % 10;
	if (mood < 3)
		return true;
	return false;
}

void VisitorWithMood::leaveQueue()
{
	this->model->addRating(3);
	this->cashDesk->removeVisitor(this);
}

File: python/extension5/VisitorWithMood.py
from __future__ import annotations
import random
from typing import TYPE_CHECKING

from python.root.Visitor import Visitor

if TYPE_CHECKING:
    from python.root.CoffeeHouse import CoffeeHouse


class VisitorWithMood(Visitor):
    def __init__(self, model: CoffeeHouse):
        super().__init__(model)
        if self._ifWantsToLeave():
            self._leaveQueue()

    def process(self, currentTime: int) -> None:
        if self._ifWantsToLeave():
            self._leaveQueue()
        else:
            super().process(currentTime)

    def _ifWantsToLeave(self) -> bool:
        mood = random.choice(range(10))
        if mood < 3:
            return True
        return False

    def _leaveQueue(self) -> None:
        self._model.addRating(3)
        self._cashDesk.removeVisitor(self)

